### PR TITLE
fileCache functionality

### PR DESCRIPTION
**Pull request by limpkin**

### CODE REVIEW

**Comment by limpkin** [comment]
@raoulh : this is ready to be merged after your review.

File: src/FilesCache.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+bool FilesCache::save(QList<QPair<int, QString>> files)
+{
+    QFile file(m_filePath);
+
+    if (!file.open(QIODevice::WriteOnly | QIODevice::Text))
+        return false;
+
+    QTextStream out(&file);
+    out << m_simpleCrypt.encryptToString(QString::number(m_dbChangeNumber)) + "\n";
+    for (QPair<int, QString> file : files) {
```
Comment (raoulh) [inline src/FilesCache.cpp]: { on a new line, please keep the existing formating of the project

File: src/FilesManagement.cpp [hunk 2]
```diff
@@ -108,6 +108,12 @@ FilesManagement::FilesManagement(QWidget *parent) :
 
     connect(ui->filesListView->selectionModel(), &QItemSelectionModel::currentChanged, this, &FilesManagement::currentSelectionChanged);
     currentSelectionChanged(QModelIndex(), QModelIndex());
+    connect(ui->listFilesButton, &QPushButton::clicked, [=](){
```
Comment (raoulh) [inline src/FilesManagement.cpp]: same here, { on a newline

File: src/FilesManagement.cpp [hunk 3]
```diff
@@ -124,6 +130,13 @@ void FilesManagement::setWsClient(WSClient *c)
         loadModel();
         ui->lineEditFilterFiles->clear();
     });
+    connect(wsClient, &WSClient::filesCacheChanged, [=]()
+    {
+        loadFilesCacheModel();
+    });
+    connect(wsClient, &WSClient::wsConnected, [=] () {
```
Comment (raoulh) [inline src/FilesManagement.cpp]: { on newline

File: src/FilesManagement.cpp [hunk 4]
```diff
@@ -170,6 +183,59 @@ void FilesManagement::loadModel()
...
+        QLabel *icon = new QLabel(listWidget);
+        icon->setPixmap(AppGui::qtAwesome()->icon(fa::fileo).pixmap(18, 18));
+        rowLayout->addWidget(icon);
+        rowLayout->addWidget(new QLabel(entry.toString()));
+
+        QToolButton *button = new QToolButton;
+        button->setToolButtonStyle(Qt::ToolButtonIconOnly);
+        button->setIcon(AppGui::qtAwesome()->icon(fa::save));
+
+        connect(button, &QToolButton::clicked, [=]() {
```
Comment (raoulh) [inline src/FilesManagement.cpp]: { on newline

File: src/FilesManagement.cpp [hunk 5]
```diff
@@ -170,6 +183,59 @@ void FilesManagement::loadModel()
...
+            fileName = QFileDialog::getSaveFileName(this, tr("Save to file..."), entry.toString());
+
+            if (fileName.isEmpty())
+                return;
+
+//            ui->progressBar->setMinimum(0);
+//            ui->progressBar->setMaximum(0);
+//            ui->progressBar->setValue(0);
+//            ui->progressBar->show();
+//            updateButtonsUI();
```
Comment (raoulh) [inline src/FilesManagement.cpp]: Remove commented line if they are not used

File: src/MPDevice.cpp [hunk 6]
```diff
@@ -263,6 +267,54 @@ void MPDevice::runAndDequeueJobs()
     currentJobs->start();
 }
 
+void MPDevice::updateFilesCache()
+{
+    getStoredFiles([=](bool success, QStringList fileNames) {
```
Comment (raoulh) [inline src/MPDevice.cpp]: { on newline

File: src/MPDevice.cpp [hunk 7]
```diff
@@ -263,6 +267,54 @@ void MPDevice::runAndDequeueJobs()
...
+        QList<QPair<int, QString>> list;
+        for (auto fileName: fileNames)
+            list.append(QPair<int, QString>(0, fileName));
+
+        if (success) {
```
Comment (raoulh) [inline src/MPDevice.cpp]: { on newline

File: src/MPDevice.cpp [hunk 8]
```diff
@@ -6125,3 +6230,66 @@ void MPDevice::importDatabase(const QByteArray &fileData, bool noDelete,
...
+        Q_UNUSED(failedJob);
+        qCritical() << "Setting device in MMM failed";
+        exitMemMgmtMode(false);
+
+        //TODO: the errCode parameter could be used as an
+        //identifier for a translated error message
+        //The string is used for client (CLI for ex.) that
+        //does not want to use the error code
+        cb(false, QStringList());
+    });
```
Comment (raoulh) [inline src/MPDevice.cpp]: Copy paste comment? That does not seems to be a valid comment for this line

**Pull request merged**